flexmeasures/data/models/data_sources.py

```diff
@@ -144,3 +168,18 @@ def to_dict(self) -> dict:
             type=self.type if self.type in ("forecaster", "scheduler") else "other",
             description=self.description,
         )
+
+    @staticmethod
+    def hash_attributes(attributes: dict) -> str:
+        return hashlib.sha256(json.dumps(attributes).encode("utf-8")).digest()
+
+    def get_attribute(self, attribute: str, default: Any = None) -> Any:
+        """Looks for the attribute on the DataSource."""
+        return self.attributes.get(attribute)
+
+    def has_attribute(self, attribute: str) -> bool:
+        return attribute in self.attributes
+
+    def set_attribute(self, attribute: str, value):
+        if self.has_attribute(attribute):
+            self.attributes[attribute] = value
```

Comment:
This effectively only allows editing an existing attribute, and not adding a new one. If that's what you were going for, the docstring should explain.

Reply:
Indeed. Changing this method to allow creating new attributes.